### PR TITLE
More EnC semantic analysis

### PR DESCRIPTION
**Pull request by tmat**
Analyze signatures, type parameters and constraints in semantic analysis.

### CODE REVIEW

**Comment by tmat** [comment]
@davidwengier PTAL. Almost all rude edits are now reported in semantic analysis.

File: src/Features/VisualBasic/Portable/EditAndContinue/SyntaxComparer.vb [hunk 1]
```diff
@@ -72,15 +72,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.EditAndContinue
 
             TypeParameterList                ' tied to parent
             TypeParameter                    ' tied to parent
-            TypeParameterConstraintClause    ' tied to parent
```
Comment (tmat) [inline src/Features/VisualBasic/Portable/EditAndContinue/SyntaxComparer.vb:75]: >TypeParameterConstraintClause [](http://example.com/codeflow?start=12&length=29)

These nodes can be treated as parts of their parent nodes - we want to see updates for the parent nodes that are associated with symbols in semantic analysis.

File: src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs [hunk 2]
```diff
@@ -11458,36 +11509,95 @@ public void FieldModifierAndTypeUpdate()
...
+                Diagnostic(RudeEditKind.Move, "G", FeaturesResources.field),
+                Diagnostic(RudeEditKind.TypeUpdate, "string G, F", FeaturesResources.field),
+                Diagnostic(RudeEditKind.TypeUpdate, "string G, F", FeaturesResources.field),
+                Diagnostic(RudeEditKind.TypeUpdate, "double V, U", FeaturesResources.field),
+                Diagnostic(RudeEditKind.Delete, "string G, F", DeletedSymbolDisplay(FeaturesResources.field, "H")));
         }
 
         [Fact]
-        public void EventFieldReorder()
+        public void EventField_Reorder()
```
Comment (davidwengier) [inline src/EditorFeatures/CSharpTest/EditAndContinue/TopLevelEditingTests.cs:11600]: Nit: Field_Event_Reorder based on the above naming

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 3]
```diff
@@ -2324,31 +2351,65 @@ public int GetHashCode(IAssemblySymbol? obj)
...
+               oldParameter.HasNotNullConstraint == newParameter.HasNotNullConstraint &&
+               oldParameter.HasUnmanagedTypeConstraint == newParameter.HasUnmanagedTypeConstraint &&
+               oldParameter.Variance == newParameter.Variance;
+
+        protected static bool TypeParametersEquivalent(ImmutableArray<ITypeParameterSymbol> oldParameters, ImmutableArray<ITypeParameterSymbol> newParameters, bool exact)
+            => oldParameters.SequenceEqual(newParameters, exact, (oldParameter, newParameter, exact) => oldParameter.Name == newParameter.Name && TypeParameterConstraintsEquivalent(oldParameter, newParameter, exact));
+
+        protected static bool BaseTypesEquivalent(INamedTypeSymbol oldType, INamedTypeSymbol newType, bool exact)
+            => TypesEquivalent(oldType.BaseType, newType.BaseType, exact) &&
+               oldType.Interfaces.SequenceEqual(newType.Interfaces, exact, (x, y, exact) => TypesEquivalent(x, y, exact));
```
Comment (davidwengier) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:2412]: Should this be AllInterfaces?
eg:
```
public interface B { }
public interface A : B { }

// source 1
public class C : A, B { }

// source2
public class C : A
```

Interfaces would be different, but AllInterfaces wouldn't be.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 4]
```diff
@@ -3231,6 +3269,7 @@ public int GetHashCode([DisallowNull] SemanticEditInfo obj)
                 => obj.Symbol.GetHashCode();
         }
 
+<<<<<<< HEAD
```
Comment (davidwengier) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Oops

**Review by davidwengier — APPROVED** [review]
:shipit:

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 3 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:2410]: Good point.

File: src/EditorFeatures/Test/Utilities/SymbolEquivalenceComparerTests.cs [hunk 5]
```diff
@@ -1118,8 +1118,8 @@ class C
             var method_v1 = type1_v1.GetMembers("M").Single();
             var method_v2 = type1_v2.GetMembers("M").Single();
 
-            var trueComp = new SymbolEquivalenceComparer(assemblyComparerOpt: null, distinguishRefFromOut: true, tupleNamesMustMatch: false);
-            var falseComp = new SymbolEquivalenceComparer(assemblyComparerOpt: null, distinguishRefFromOut: false, tupleNamesMustMatch: false);
+            var trueComp = new SymbolEquivalenceComparer(assemblyComparerOpt: null, distinguishRefFromOut: true, tupleNamesMustMatch: false, ignoreNullableAnnotations: false);
+            var falseComp = new SymbolEquivalenceComparer(assemblyComparerOpt: null, distinguishRefFromOut: false, tupleNamesMustMatch: false, ignoreNullableAnnotations: false);
```
Comment (sharwell) [inline src/EditorFeatures/Test/Utilities/SymbolEquivalenceComparerTests.cs:1122]: 😕 What's the difference between this a `SymbolEqualityComparer.Default`?

Comment (tmat) [reply src/EditorFeatures/Test/Utilities/SymbolEquivalenceComparerTests.cs:1122]: The test is testing `distinguishRefFromOut` parameter, so it creates the comparer explicitly with that parameter set to true/false.

File: src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs [hunk 6]
```diff
@@ -1331,12 +1328,18 @@ static bool DiffersInReadOnlyModifier(SyntaxTokenList oldModifiers, SyntaxTokenL
...
+                    // Inserting/deleting a type parameter constraint should result in an update of the corresponding type parameter symbol:
                     if (node.IsKind(SyntaxKind.TypeParameterConstraintClause))
                     {
                         return OneOrMany.Create(ImmutableArray.Create((oldSymbol, newSymbol, EditKind.Update)));
                     }
 
+                    // Inserting/deleting a global statement should result in an update of the implicit main method:
+                    if (node.IsKind(SyntaxKind.GlobalStatement))
+                    {
+                        return OneOrMany.Create(ImmutableArray.Create((oldSymbol, newSymbol, EditKind.Update)));
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/EditAndContinue/CSharpEditAndContinueAnalyzer.cs:1340]: 👍

I like this method of changing the edit kind in `GetSymbols`

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 7]
```diff
@@ -3477,8 +3394,9 @@ private static void ReportGlobalMainMethodRudeEdits(IMethodSymbol oldSymbol, IMe
 
             if (rudeEdit != RudeEditKind.None)
             {
-                var arguments = (rudeEdit == RudeEditKind.TypeKindUpdate) ? Array.Empty<string>() : null;
-                ReportUpdateRudeEdit(diagnostics, rudeEdit, oldSymbol, newSymbol, newNode, newCompilation, cancellationToken, arguments);
+                // so we'll just use the last global statement in the file
```
Comment (davidwengier) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3397]: Nit: this comment looks out of place

Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:3397]: Will clean up in a follow up PR.

**Pull request merged**